### PR TITLE
Created test that validates whether a background coda is spun up

### PR DESCRIPTION
**Pull request by wu-s-john**
Checklist:

- [X] Tests were added for the new behavior
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Does this close issues? List them:

Closes #0000
Closes #0000

### CODE REVIEW

File: src/lib/command_line_tests/command_line_tests.ml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+          ~args:
+            [ "exec"
+            ; "coda"
+            ; "daemon"
+            ; "--"
+            ; "-background"
+            ; "-client-port"
+            ; sprintf "%d" port ]
+          ()
+      in
```
Comment (cmr) [inline src/lib/command_line_tests/command_line_tests.ml]: Put a sleep here to make sure the daemon gets a chance to initialize. 100ms should be plenty.

File: src/lib/command_line_tests/command_line_tests.ml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            [ "exec"
+            ; "coda"
+            ; "client"
+            ; "status"
+            ; "--"
+            ; "-daemon-port"
+            ; sprintf "%d" port ]
+          ()
+      in
+      let%map _ : string = kill daemon_process in
```
Comment (cmr) [inline src/lib/command_line_tests/command_line_tests.ml]: use `coda client stop`

File: src/lib/command_line_tests/command_line_tests.ml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+let%test_module "Command line tests" =
+  ( module struct
+    let kill port =
+      Process.run_exn () ~prog:"dune"
+        ~args:
+          [ "exec"
+          ; "coda"
+          ; "client"
+          ; "stop"
```
Comment (cmr) [inline src/lib/command_line_tests/command_line_tests.ml]: stop-daemon

File: src/lib/command_line_tests/command_line_tests.ml [hunk 4]
```diff
@@ -0,0 +1,56 @@
+open Core
+open Async
+
+(**
+ * This module tests the basic funtionalities of coda through the cli
```
Comment (psteckler) [inline src/lib/command_line_tests/command_line_tests.ml]: functionality

File: src/lib/command_line_tests/command_line_tests.ml (hunk 1 above)
Comment (psteckler) [reply src/lib/command_line_tests/command_line_tests.ml]: Is 100 milliseconds still enough in the presence of the hash chunk table deserialization, which occurs (only) in the daemon?

That deserialization may take 3 seconds.

**Comment by cmr** [comment]
This cannot be run with `runtest` until after we have already built coda. Otherwise the recursive `dune exec` will fail or try and build another coda in a new directory which is definitely wrong. In CI this is fine I think, but on my local machine at least I often use just a bare `dune runtest` and expect it to do my build. 

Adding a `--root` argument pointing at the repo root (or src? we build in src I think) will fix the current failure that it can't find a `coda` executable.

**Comment by jkrauska** [comment]
Seeing if we can revive this.

**Comment by psteckler** [comment]
There's a temporary intentional delay when starting the daemon, the amount depends on the profile. It will be removed soon, according to @ghost-not-in-the-shell. 

When that happens, we can reduce the client wait delay in this PR.

File: src/lib/command_line_tests/command_line_tests.ml [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        ()
+
+    let create_config_directory config_dir =
+      (* create empty config dir to avoid any issues with the default config dir *)
+      let%bind _ = Process.run_exn ~prog:"rm" ~args:["-rf"; config_dir] () in
+      let%map _ = Process.run_exn ~prog:"mkdir" ~args:["-p"; config_dir] () in
+      ()
+
+    let test_background_daemon () =
+      let config_dir = "/tmp/coda-spun-test" in
```
Comment (cmr) [inline src/lib/command_line_tests/command_line_tests.ml]: Can you use `File_system.with_temp_dir` for this?

File: src/lib/command_line_tests/command_line_tests.ml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+ *)
+
+let%test_module "Command line tests" =
+  ( module struct
+    (* executable location relative to src/default/lib/command_line_tests
+
+       dune won't allow running it via "dune exec", because it's outside its
+       workspace, so we invoke the executable directly
+
+       the coda.exe executable must have been built before running the test
```
Comment (cmr) [inline src/lib/command_line_tests/command_line_tests.ml]: Can we add an explicit test for the existence of `coda.exe`, with an error message to run `dune build`? I just want to be friendlier to wrong uses of `dune runtest`.

File: src/lib/command_line_tests/dune [hunk 7]
```diff
@@ -0,0 +1,7 @@
+(library
+  (name command_line_tests)
+  (public_name command_line_tests)
+  (inline_tests)
+  (libraries async)
```
Comment (cmr) [inline src/lib/command_line_tests/dune]: Or, maybe we can somehow make this depend on `coda.exe` and dune will make sure it is built for us? Is it possible to depend on an executable?

Comment (psteckler) [reply src/lib/command_line_tests/dune]: I looked through the `dune` docs for that, didn't find anything. I can print a message and fail if the executable's not there.

File: src/lib/command_line_tests/command_line_tests.ml (hunk 5 above)
Comment (psteckler) [reply src/lib/command_line_tests/command_line_tests.ml]: Yes.

**Pull request merged**